lib/modules/manager/pip_requirements/artifacts.ts

```diff
@@ -25,10 +28,13 @@ export async function updateArtifacts({
       .split(newlineRegex)
       .map((line) => line.trim());
     for (const dep of updatedDeps) {
+      const depName = dep.depName!;
       const hashLine = lines.find(
         (line) =>
           // TODO: types (#7154)
-          line.startsWith(`${dep.depName!}==`) && line.includes('--hash=')
+          line.startsWith(depName) &&
+          extrasAndEqualsRegex.test(line.substring(depName.length)) &&
+          line.includes('--hash=')
```

Comment:
instead of these 3 lines, can we create 1 regex that supports all lines?

Reply:
The `line.includes('--hash=')` won't be hard to cover by some regex. The dependency name is more problematic – it's a string we do not control and we either need to be 100% sure it does not contain any special characters or we need to escape it in the regex.

I am everything but a JavaScript or TypeScript expert. AFAIK – which boils down to the [regex chapter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) in the MSDN JavaScript guide – the JavaScript regex engine does not support the `\Q`/`\E` quotation escapes or a function like Python's `re.escape()`.  Is there some helper function in the renovate codebase that escapes a string so it matches exactly the input string and nothing else?

Besides, keeping the test for the dependency name separate means the regular expression needs to be compiled only once, not for every single dependency. I don't expect this will have a huge impact in the end, though.

Regarding `lines.find( (line) => ...`:  I don't think it is needed in a strict sense. We certainly could come up with a regex that also accounts for the whitespace [trimmed from each line](https://github.com/renovatebot/renovate/blob/0b0fbd6ee81965ad652c69b1ff9f80fc759e010f/lib/modules/manager/pip_requirements/artifacts.ts#L29) and make sure the regex matches within single lines only. I am not sure whether the resulting code would be easier to understand, though. (The majority of the developers I know consider regular expressions hard to read.) And, of course, I did not want to make more changes to the codebase than necessary.

Reply:
resolved in d540d0cf1fdb3e2621920edc50790f38ad65150e

Reply:
Ok, I found `util.regex.escapeRegExp()`. I will give the single regex test a try.

Reply:
It's in e8eccbf48099da83b0b6cc15b441160e363fc65e. You can decide whether you prefer the previous version or this one that requires the reader to think a bit more about the regexp.